# Training of Recurrenct Neural Network (RNN) model on spectrograms ( second version of spectrograms)

In [1]:
import numpy as np
import pandas as pd

from scipy import signal
from scipy.io import wavfile
import matplotlib.pyplot as plt

from PIL import Image
import matplotlib
import random

In [2]:
import tensorflow as tf
from tensorflow import keras
from keras import layers


tf.config.run_functions_eagerly

<function tensorflow.python.eager.polymorphic_function.polymorphic_function.run_functions_eagerly(run_eagerly)>

In [3]:
# Loading of datasets

spectr_train_list = pd.read_csv("spectr2_train_list.txt", index_col = 0).iloc[:,0].values.tolist()
spectr_test_list = pd.read_csv("spectr2_test_list.txt", index_col = 0).iloc[:,0].values.tolist()

# labels_train, labels_test
# spectrograms_train, spectrograms_test

labels_train = []
spectrograms_train = []

for path in spectr_train_list:
    labels_train.append(path.split("/")[1])
    
    
    rgb = matplotlib.image.imread(path)
    gray = np.mean(rgb, -1)
    spectrogram = (255*(gray - np.min(gray))/np.ptp(gray)).astype(int)  # [0, 255] scaled

    spectrograms_train.append(spectrogram)

labels_test = []
spectrograms_test = []

for path in spectr_test_list:
    labels_test.append(path.split("/")[1])
    
    rgb = matplotlib.image.imread(path)
    gray = np.mean(rgb, -1)
    spectrogram = (255*(gray - np.min(gray))/np.ptp(gray)).astype(int)  # [0, 255] scaled
    
    spectrograms_test.append(spectrogram)


C:\Users\patry\AppData\Local\Temp\ipykernel_18340\3981592614.py:18: RuntimeWarning: invalid value encountered in divide
  spectrogram = (255*(gray - np.min(gray))/np.ptp(gray)).astype(int)  # [0, 255] scaled
C:\Users\patry\AppData\Local\Temp\ipykernel_18340\3981592614.py:30: RuntimeWarning: invalid value encountered in divide
  spectrogram = (255*(gray - np.min(gray))/np.ptp(gray)).astype(int)  # [0, 255] scaled


In [4]:
labels_test[0:9]

['no', 'up', 'off', 'go', 'yes', 'stop', 'down', 'on', 'down']

In [8]:
# preparing datasets for models

dummies = pd.get_dummies(labels_train)
y_train = tf.one_hot(dummies,10)

dataset_train = tf.data.Dataset.from_tensor_slices((spectrograms_train, dummies)).shuffle(200).batch(8)

####################

dummies2 = pd.get_dummies(labels_test)
y_test = tf.one_hot(dummies2, 10)

dataset_test = tf.data.Dataset.from_tensor_slices((spectrograms_test, dummies2)).batch(8)

#####################

early_stop = tf.keras.callbacks.EarlyStopping(monitor='val_accuracy',
                           mode='max',
                           verbose=1,
                           patience=4)


# Model architecture

In [12]:
# model - LSTM
def get_model_RNN():
  model = tf.keras.models.Sequential()

  model.add(tf.keras.layers.LSTM(units = 64, return_sequences = True, input_shape=[64,64]))
  model.add(tf.keras.layers.LSTM(units = 64, return_sequences = True))

  model.add(tf.keras.layers.Dense(units = 32))
  model.add(tf.keras.layers.Flatten())
  model.add(tf.keras.layers.Dense(units = 10, activation='softmax'))

  model.compile(loss = 'categorical_crossentropy',
             optimizer = keras.optimizers.Adam(learning_rate = 0.0001),
             metrics = ['accuracy'])
  return model

# Model training & ealuation

In [63]:
# model 1 ( using spectrograms v1 - for comparison)

model_RNN = get_model_RNN()

model_RNN.fit(dataset_train, epochs=50,  callbacks =[early_stop], validation_data=dataset_test)

Epoch 1/50
483/483 [==============================] - 32s 55ms/step - loss: 2.2395 - accuracy: 0.1606 - val_loss: 2.1018 - val_accuracy: 0.2665
Epoch 2/50
483/483 [==============================] - 25s 52ms/step - loss: 1.9484 - accuracy: 0.3098 - val_loss: 1.7458 - val_accuracy: 0.4033
Epoch 3/50
483/483 [==============================] - 25s 51ms/step - loss: 1.6072 - accuracy: 0.4389 - val_loss: 1.4957 - val_accuracy: 0.4822
Epoch 4/50
483/483 [==============================] - 24s 50ms/step - loss: 1.3887 - accuracy: 0.5109 - val_loss: 1.3376 - val_accuracy: 0.5482
Epoch 5/50
483/483 [==============================] - 24s 50ms/step - loss: 1.2319 - accuracy: 0.5736 - val_loss: 1.2266 - val_accuracy: 0.5769
Epoch 6/50
483/483 [==============================] - 25s 51ms/step - loss: 1.1074 - accuracy: 0.6197 - val_loss: 1.1362 - val_accuracy: 0.6148
Epoch 7/50
483/483 [==============================] - 25s 51ms/step - loss: 0.9966 - accuracy: 0.6554 - val_loss: 1.0613 - val_accuracy:

In [9]:
# model 1 - using spectrograms v2 (after logarithmic scaling)

model_RNN = get_model_RNN()

model_RNN.fit(dataset_train, epochs=50,  callbacks =[early_stop], validation_data=dataset_test)

Epoch 1/50
483/483 [==============================] - 37s 66ms/step - loss: 2.3212 - accuracy: 0.1096 - val_loss: 2.3000 - val_accuracy: 0.1175
Epoch 2/50
483/483 [==============================] - 30s 63ms/step - loss: 2.2976 - accuracy: 0.1205 - val_loss: 2.2893 - val_accuracy: 0.1333
Epoch 3/50
483/483 [==============================] - 30s 62ms/step - loss: 2.2801 - accuracy: 0.1448 - val_loss: 2.2817 - val_accuracy: 0.1449
Epoch 4/50
483/483 [==============================] - 30s 61ms/step - loss: 2.2634 - accuracy: 0.1513 - val_loss: 2.2748 - val_accuracy: 0.1520
Epoch 5/50
483/483 [==============================] - 29s 60ms/step - loss: 2.2490 - accuracy: 0.1699 - val_loss: 2.2635 - val_accuracy: 0.1607
Epoch 6/50
483/483 [==============================] - 30s 61ms/step - loss: 2.2302 - accuracy: 0.1894 - val_loss: 2.2583 - val_accuracy: 0.1654
Epoch 7/50
483/483 [==============================] - 30s 63ms/step - loss: 2.2105 - accuracy: 0.1940 - val_loss: 2.2465 - val_accuracy:

Accuracy: 0.45\%

# Model 2 - more robust

In [14]:
# model - LSTM
def get_model_RNN2():
  model = tf.keras.models.Sequential()

  model.add(tf.keras.layers.LSTM(units = 128, return_sequences = True, input_shape=[64,64]))
  model.add(tf.keras.layers.LSTM(units = 128, return_sequences = True))

  model.add(tf.keras.layers.Dense(units = 64))
  model.add(tf.keras.layers.Flatten())
  model.add(tf.keras.layers.Dense(units = 10, activation='softmax'))

  model.compile(loss = 'categorical_crossentropy',
             optimizer = keras.optimizers.Adam(learning_rate = 0.0001),
             metrics = ['accuracy'])
  return model

In [65]:
# model 2

model_RNN2 = get_model_RNN2()

model_RNN2.fit(dataset_train, epochs=50,  callbacks =[early_stop], validation_data=dataset_test)

Epoch 1/50
483/483 [==============================] - 49s 91ms/step - loss: 2.1186 - accuracy: 0.2306 - val_loss: 1.7894 - val_accuracy: 0.3892
Epoch 2/50
483/483 [==============================] - 43s 90ms/step - loss: 1.6071 - accuracy: 0.4293 - val_loss: 1.4027 - val_accuracy: 0.5237
Epoch 3/50
483/483 [==============================] - 45s 92ms/step - loss: 1.2763 - accuracy: 0.5459 - val_loss: 1.1683 - val_accuracy: 0.5891
Epoch 4/50
483/483 [==============================] - 45s 93ms/step - loss: 1.0340 - accuracy: 0.6324 - val_loss: 1.0353 - val_accuracy: 0.6452
Epoch 5/50
483/483 [==============================] - 44s 91ms/step - loss: 0.8422 - accuracy: 0.7000 - val_loss: 0.9232 - val_accuracy: 0.6739
Epoch 6/50
483/483 [==============================] - 45s 93ms/step - loss: 0.6917 - accuracy: 0.7518 - val_loss: 0.8720 - val_accuracy: 0.7089
Epoch 7/50
483/483 [==============================] - 44s 90ms/step - loss: 0.5552 - accuracy: 0.8070 - val_loss: 0.8621 - val_accuracy:

In [67]:
model_RNN2.fit(dataset_train, epochs=20,  callbacks =[early_stop], validation_data=dataset_test)

Epoch 1/20
483/483 [==============================] - 44s 91ms/step - loss: 0.1433 - accuracy: 0.9617 - val_loss: 0.9758 - val_accuracy: 0.7797
Epoch 2/20
483/483 [==============================] - 44s 90ms/step - loss: 0.0965 - accuracy: 0.9744 - val_loss: 1.1706 - val_accuracy: 0.7621
Epoch 3/20
483/483 [==============================] - 45s 93ms/step - loss: 0.0722 - accuracy: 0.9821 - val_loss: 1.1034 - val_accuracy: 0.7855
Epoch 4/20
483/483 [==============================] - 43s 89ms/step - loss: 0.0773 - accuracy: 0.9790 - val_loss: 1.0642 - val_accuracy: 0.7914
Epoch 5/20
483/483 [==============================] - 43s 89ms/step - loss: 0.0562 - accuracy: 0.9876 - val_loss: 1.1104 - val_accuracy: 0.7943
Epoch 6/20
483/483 [==============================] - 44s 90ms/step - loss: 0.0474 - accuracy: 0.9896 - val_loss: 1.1606 - val_accuracy: 0.7995
Epoch 6: early stopping


In [15]:
# model 2 more robust - using spectrograms v2 (after logarithmic scaling)

# model 2

model_RNN2 = get_model_RNN2()

model_RNN2.fit(dataset_train, epochs=50,  callbacks =[early_stop], validation_data=dataset_test)


Epoch 1/50
483/483 [==============================] - 24s 45ms/step - loss: 2.3235 - accuracy: 0.1016 - val_loss: 2.2914 - val_accuracy: 0.1257
Epoch 2/50
483/483 [==============================] - 22s 45ms/step - loss: 2.2832 - accuracy: 0.1438 - val_loss: 2.2652 - val_accuracy: 0.1613
Epoch 3/50
483/483 [==============================] - 22s 45ms/step - loss: 2.2584 - accuracy: 0.1653 - val_loss: 2.2481 - val_accuracy: 0.1730
Epoch 4/50
483/483 [==============================] - 21s 44ms/step - loss: 2.2317 - accuracy: 0.1889 - val_loss: 2.2363 - val_accuracy: 0.1882
Epoch 5/50
483/483 [==============================] - 22s 45ms/step - loss: 2.1941 - accuracy: 0.2041 - val_loss: 2.2147 - val_accuracy: 0.1975
Epoch 6/50
483/483 [==============================] - 22s 46ms/step - loss: 2.1561 - accuracy: 0.2290 - val_loss: 2.2090 - val_accuracy: 0.2016
Epoch 7/50
483/483 [==============================] - 22s 45ms/step - loss: 2.0906 - accuracy: 0.2562 - val_loss: 2.1748 - val_accuracy:

Accuracy: 0.51\%

# (old) Conclusions from training on old spectrograms:


The first model obtained **74\% accuracy** on validation data set, whereas the second model obtained almost **80\% accuracy** with the same spectrograms, optimizer and learning rate settings. 

Reminder: it was a 10 class classification problem with balanced classes.

Possible ideas to train better version of the final model:

- decreasing learning rate, as models started to overfit ( extended learning process)
- use more layers and units within each layer ( more robust models)
- prepare other version of spectrograms (better data means better model)

# Conclusions from training on new spectrograms ( v2):


The first model obtained **45\% accuracy** on validation data set, whereas the second model obtained almost **52\% accuracy** with the same spectrograms, optimizer and learning rate settings. 

Having considered this modification to the machine learning pipeline:

- prepare other version of spectrograms (better data means better model)


The models' training time was considerably longer, so the modification consisting of **creation of new version of spectrograms** -- preserving more information, **was**, in this case, **leading to worse results**. Perhaps it was caused by too much background noise in the audio files and the old version of spectrograms was sufficient.